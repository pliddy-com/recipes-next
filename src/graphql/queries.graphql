# Query for Homepage RecipeCollection (all recipes)

query homePageQuery {
  recipeCollection(order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe collection based on filter

query recipeCollection($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    total
    items {
      ...RecipeDefault
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)

query listPageQuery($where: TagFilter) {
  tagCollection(where: $where, limit: 1) {
    total
    items {
      slug
      title
      linkedFrom {
        recipeCollection(limit: 24) {
          total
          items {
            ...RecipeSummary
          }
        }
      }
    }
  }
}

# Query for recipe slugs

query recipeSlugsCollection($where: RecipeFilter) {
  recipeCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query categorySlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query tagSlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
      linkedFrom {
        recipeCollection {
          total
        }
      }
    }
  }
}

# query for Taxonomy with Category & Subcategory levels

query taxonomyCollection($where: TaxonomyFilter) {
  taxonomyCollection(where: $where, limit: 1) {
    items {
      ...TaxonomyDefault

      tag {
        ...TagDefault
      }

      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            ...TagDefault

            linkedFrom {
              recipeCollection(limit: 1) {
                total
              }
            }
          }
          ...TaxonomyWithChildren
        }
      }
    }
  }
}
