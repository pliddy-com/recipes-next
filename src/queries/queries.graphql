# query for Taxonomy with Category & Subcategory levels
# getNavTaxonomy

query navTaxonomy($where: TaxonomyFilter) {
  taxonomyCollection(where: $where, order: title_ASC, limit: 1) {
    __typename
    items {
      ...TaxonomyDefault

      tag {
        ...TagDefault
      }

      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            ...TagDefault

            linkedFrom {
              recipeCollection(limit: 1) {
                total
              }
            }
          }
          ...TaxonomyWithChildren
        }
      }
    }
  }
}

# Query for recipe slugs
# getRecipeSlugs

query recipeSlugs($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    __typename
    items {
      slug
    }
  }
}

# query for category slugs
# getTagSlugs

query tagSlugs($where: TagFilter) {
  tagCollection(where: $where, order: title_ASC) {
    __typename
    items {
      slug
      linkedFrom {
        recipeCollection {
          total
        }
      }
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)
# getRecipeList

query recipeList($where: TagFilter) {
  __typename
  tagCollection(where: $where, order: title_ASC, limit: 1) {
    total
    items {
      ...ListPageItem
    }
  }
}

# Query for Recipe Collection for home (index) page
# getRecipeIndex

query recipeIndex($where: RecipeFilter) {
  __typename
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page
# getRecipePage

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    __typename
    items {
      ...RecipeDefault
    }
  }

  categoriesTaxonomy: taxonomyCollection(
    where: { slug: "categories" }
    order: title_ASC
    limit: 1
  ) {
    __typename
    items {
      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            __typename
            title
            slug
          }
          ... on Taxonomy {
            ...TaxonomyDefault
          }
        }
      }
    }
  }

  cuisineTaxonomy: taxonomyCollection(
    where: { slug: "cuisine" }
    order: title_ASC
    limit: 1
  ) {
    items {
      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            __typename
            title
            slug
          }
          ... on Taxonomy {
            __typename
            title
            slug
          }
        }
      }
    }
  }
}

#Query for Tag Index page
# getTagIndex

query tagIndex($where: TagFilter) {
  tagCollection(where: $where, order: title_ASC) {
    __typename
    total
    items {
      ...ListPageItem
    }
  }
}
