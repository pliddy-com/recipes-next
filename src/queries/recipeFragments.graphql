# Fragments with default fields for Recipe-specific sections:
#   * Ingredients
#   * Instructions
#   * Image
#   * RecipeDefault (full)
#   * RecipeSummary (card)

fragment IngredientsDefault on IngredientSection {
  sys {
    id
  }
  title
  slug
  label
  ingredientList
}

fragment InstructionsDefault on InstructionSection {
  sys {
    id
  }
  title
  slug
  label
  instructionList
}

fragment ImageDefault on Asset {
  sys {
    id
  }
  __typename
  title
  description
  contentType
  fileName
  size
  url
  height
  width
}

fragment DescriptionDefault on Recipe {
  description {
    json
    links {
      assets {
        block {
          sys {
            id
          }
          __typename
          title
          url
        }
        hyperlink {
          sys {
            id
          }
          __typename
          title
          url
        }
      }
      entries {
        inline {
          __typename
          sys {
            id
          }
          ... on Recipe {
            slug
            title
          }
        }
        hyperlink {
          __typename
          sys {
            id
          }
        }
      }
    }
  }
}

fragment RecipeDefault on Recipe {
  sys {
    id
  }
  __typename
  title
  slug
  ...DescriptionDefault
  abstract
  image {
    ...ImageDefault
  }
  ingredientsCollection(limit: 10) {
    items {
      ...IngredientsDefault
    }
  }
  equipment
  instructionsCollection(limit: 10) {
    items {
      ...InstructionsDefault
    }
  }
  notes
  tagsCollection(limit: 10) {
    items {
      ...TagDefault
    }
  }
}

fragment RecipeSummary on Recipe {
  sys {
    id
  }
  __typename
  title
  slug
  abstract
  image {
    ...ImageDefault
  }
  tagsCollection(limit: 10) {
    items {
      ...TagDefault
    }
  }
}
