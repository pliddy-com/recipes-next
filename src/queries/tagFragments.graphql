# Fragments with default fields for:
#   * Tag
#   * Taxonomy

fragment TagDefault on Tag {
  sys {
    id
  }
  __typename
  title
  slug
}

fragment TaxonomyDefault on Taxonomy {
  sys {
    id
  }
  __typename
  title
  slug
}

# Fragment with fields for Tag with linkedFrom Recipes

fragment TagWithLinks on Tag {
  # ...TagDefault
  sys {
    id
  }
  __typename
  title
  slug

  linkedFrom {
    recipeCollection(limit: 1) {
      total
    }
  }
}

# Fragment with Fields for Taxonomy with no children

fragment TaxonomyRoot on Taxonomy {
  # ...TaxonomyDefault
  sys {
    id
  }
  __typename
  title
  slug

  tag {
    # ...TagDefault
    sys {
      id
    }
    __typename
    title
    slug
  }
}

# Fragment with Fields for Taxonomy with children

fragment TaxonomyWithChildren on Taxonomy {
  # ...TaxonomyDefault
  sys {
    id
  }
  __typename
  title
  slug

  tag {
    ...TagWithLinks
  }
  childrenCollection {
    total
    items {
      ...TagWithLinks
      ...TaxonomyRoot
    }
  }
}

# fragment for items on list page queries
fragment ListPageItem on Tag {
  slug
  title
  linkedFrom {
    recipeCollection(limit: 24) {
      total
      items {
        ...RecipeSummary
      }
    }
  }
}
