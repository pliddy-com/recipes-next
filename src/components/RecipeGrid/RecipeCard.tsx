import Link from 'next/link';

import Card from '@mui/material/Card';
import CardActionArea from '@mui/material/CardActionArea';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';

import TagButtons from '@/components/TagButtons/TagButtons';
import DynamicImage from '@/components/DynamicImage/DynamicImage';

import { RecipeDefaultFragment } from 'types/generated/graphql';

const styles = {
  abstract: {
    mb: '2rem',
  },
  wrapper: {
    height: '100%',
    position: 'relative',
    width: '100%',
  },
};

interface RecipeCardProps {
  recipe?: RecipeDefaultFragment;
}

const RecipeCard = ({ recipe }: RecipeCardProps) => {
  if (!recipe) return null;

  const { title, abstract, image, tagsCollection, slug } = recipe ?? {};
  const { items: tags } = tagsCollection ?? {};
  const { title: category } = tags?.[0] ?? {};

  return (
    <Card variant="outlined">
      <CardActionArea component={Link} href={`/recipe/${slug}`}>
        <CardHeader title={title} subheader={category} />
        {image && (
          <CardMedia sx={styles.wrapper}>
            <DynamicImage image={image} />
          </CardMedia>
        )}

        <CardContent>
          {abstract && (
            <Typography variant="body2" sx={styles.abstract}>
              {abstract}
            </Typography>
          )}
        </CardContent>
      </CardActionArea>
      <CardActions>{tags && <TagButtons tags={tags} />}</CardActions>
    </Card>
  );
};

export default RecipeCard;
