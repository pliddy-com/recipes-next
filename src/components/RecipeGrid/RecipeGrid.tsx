import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';

import Loading from 'components/Loading/Loading';
import RecipeCard from 'components/RecipeCard/RecipeCard';

import { RecipeDefaultFragment } from 'types/generated/graphql';

import { flex } from 'lib/styles';

interface RecipeGridrops {
  recipes: (RecipeDefaultFragment | null)[];
  title?: string | null;
}

const RecipeGrid = ({ recipes, title }: RecipeGridrops) =>
  recipes && recipes.length > 0 ? (
    <Container className="page">
      <Typography variant="h1">{title}</Typography>
      <Typography variant="subtitle1" component="h2" gutterBottom>
        {recipes && `${recipes.length} Total`}
      </Typography>
      <Grid container spacing={2}>
        {recipes &&
          recipes.map((recipe, index) => (
            <Grid item md={6} sm={6} xs={12} key={recipe?.slug} sx={{ flex }}>
              {recipe && <RecipeCard recipe={recipe} preloadImg={index < 6} />}
            </Grid>
          ))}
      </Grid>
    </Container>
  ) : (
    <Loading />
  );

export default RecipeGrid;
