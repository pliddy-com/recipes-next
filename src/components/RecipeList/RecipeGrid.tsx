import Grid from '@mui/material/Grid';

import Loading from 'components/Loading';
import RecipeCard from 'components/RecipeList/RecipeCard';

import { Maybe, RecipeDefaultFragment } from 'types/generated/graphql';

type RecipeGridProps = {
  recipes?: RecipeDefaultFragment[];
};

const styles = {
  grid: {
    display: 'flex',
  },
};

const RecipeGrid = ({ recipes }: RecipeGridProps) => {
  if (!recipes) return <Loading />;

  return (
    <Grid container spacing={2}>
      {recipes &&
        recipes.map((recipe: Maybe<RecipeDefaultFragment>) => (
          <Grid item md={4} sm={6} xs={12} key={recipe?.slug} sx={styles.grid}>
            <RecipeCard recipe={recipe} />
          </Grid>
        ))}
    </Grid>
  );
};

export default RecipeGrid;
