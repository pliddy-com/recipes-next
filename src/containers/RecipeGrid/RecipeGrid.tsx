import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';

import Loading from 'components/Loading/Loading';
import RecipeCard from 'containers/RecipeGrid/RecipeCard';

import { RecipeDefaultFragment } from 'types/generated/graphql';

interface RecipeGridrops {
  recipes: (RecipeDefaultFragment | null)[];
  title?: string | null;
}

const styles = {
  gridItem: {
    display: 'flex',
  },
};

const RecipeGrid = ({ recipes, title }: RecipeGridrops) => {
  if (!recipes) return <Loading />;

  return (
    <Container className="page">
      <Typography variant="h1">{title}</Typography>
      <Typography variant="subtitle1" component="h2" gutterBottom>
        {recipes && `${recipes.length} Total`}
      </Typography>
      <Grid container spacing={2}>
        {recipes &&
          recipes.map((recipe) => (
            <Grid
              item
              md={4}
              sm={6}
              xs={12}
              key={recipe?.slug}
              sx={styles.gridItem}
            >
              {recipe && <RecipeCard recipe={recipe} />}
            </Grid>
          ))}
      </Grid>
    </Container>
  );
};

export default RecipeGrid;
