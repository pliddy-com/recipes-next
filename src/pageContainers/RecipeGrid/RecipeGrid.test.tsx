// import testing-library methods
import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

// import the component to test
import RecipeGrid from './RecipeGrid';

import { RecipeDefaultFragment } from 'types/queries';

describe('RecipeGrid', () => {
  describe('when there is page content', () => {
    it('it renders the RecipeGrid if there is content', () => {
      const recipes: (RecipeDefaultFragment | null)[] = [
        {
          sys: {
            id: 'sysid-0',
            __typename: 'Sys',
          },
          __typename: 'Recipe',
          title: 'Title 1',
          slug: 'slug-1',
          tagsCollection: {
            items: [
              {
                sys: {
                  id: 'sysid-1',
                  __typename: 'Sys',
                },
                __typename: 'Tag',
                title: 'Title 2',
                slug: 'slug-2',
              },
            ],
            __typename: 'RecipeTagsCollection',
          },
        },
      ];

      const title = 'Title';

      const { container } = render(
        <RecipeGrid recipes={recipes} title={title} />
      );

      // assert that page container is present
      const page = document.querySelector('.page');
      expect(page).toBeInTheDocument();

      // assert that card component is rendered
      const card = document.querySelector('.MuiCard-root');
      expect(card).toBeInTheDocument();

      // assert that the component matches the existing snapshot
      expect(container).toMatchSnapshot();
    });
  });

  describe('when there is no pageContent', () => {
    it('it does not render the RecipeGrid', () => {
      const title = 'Title';
      const recipes = [] as RecipeDefaultFragment[];

      render(<RecipeGrid title={title} recipes={recipes} />);

      // test if card compoent is not rendered
      const card = document.querySelector('.MuiCard-root');
      expect(card).toBeNull();
    });
  });
});
