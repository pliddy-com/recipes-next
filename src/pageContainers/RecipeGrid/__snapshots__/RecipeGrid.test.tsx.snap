// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeGrid when there is page content it renders the RecipeGrid if there is content 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg page css-1oqqzyl-MuiContainer-root"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
    >
      Title
    </h1>
    <h2
      class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-1wniyei-MuiTypography-root"
    >
      1 Total
    </h2>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 css-wq3jbm-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root recipe css-1v5z4dq-MuiPaper-root-MuiCard-root"
        >
          <a
            class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
            href="/recipe/slug-1"
            tabindex="0"
          >
            <div
              class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
                >
                  Title 1
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiCardHeader-subheader css-nrdprl-MuiTypography-root"
                >
                  Title 2
                </span>
              </div>
            </div>
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            />
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <div
              class="css-180l46g-MuiStack-root"
              data-testid="tag-buttons"
            >
              <a
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-wmndcn-MuiButtonBase-root-MuiButton-root"
                href="/tag/slug-2"
                role="button"
                tabindex="0"
              >
                 Title 2 
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
